### ISSUE TITLE
How to deal with the nan of regularizer_loss?

### ISSUE THREAD

**Issue by Xiaocong6** [google-research/morph-net#70]
Hi, thanks for open soure.
I met  a problem: the regularizer_loss is nan
the code is resnet50 in tensorflow
the goal is classification of cifar-100
the loss used is GammaFlopsRegularizer
I tried multiple regularization_strength, but it is still nan.
What should I do to solve this problem? Thanks for your answer.

**Comment by pkch** [comment]
Could you share the full code you used to run your experiment?

**Comment by eladeban** [comment]
In particular:

1. What regularization strength are you trying?
2. Could you try 1e-12 just as a sanity check? 
3. what framework are you using layers.contrib? keras?
4. is the regularization loss NaNs out immediately or dose it start at some value and then goes up until the NaN?

**Comment by Xiaocong6** [comment]
Thanks, I am sorry that I did not make it clear.
I tried  regularization strength from 1e-12 to 1e-8.
The regularization loss is NaNs out immediately.
![12](https://user-images.githubusercontent.com/44869509/58063721-ef8e0280-7bb1-11e9-9bd1-ead045fb17b7.png)


The code in train.py is as follow:
`x_input = tf.placeholder(tf.float32, [None, 32, 32, 3])
 y_input = tf.placeholder(tf.int64, [None, ])
 lr = tf.placeholder(tf.float32, [])
 if network == 'resnet50':
     prob = resnet50(x_input, is_training=True, kernel_initializer=tf.orthogonal_initializer())
 network_regularizer = flop_regularizer.GammaFlopsRegularizer([prob.op], gamma_threshold=1e-3)
 regularization_strength = 1e-12
 regularizer_loss = (network_regularizer.get_regularization_term() * regularization_strength)
 
 loss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(logits=prob, labels=y_input))
 if opt == 'momentum':
     opt = tf.train.MomentumOptimizer(lr, 0.9)

 update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
 with tf.control_dependencies(update_ops):
     train_op = opt.minimize(loss+regularizer_loss)
 logit_softmax = tf.nn.softmax(prob)
 acc  = tf.reduce_mean(tf.cast(tf.equal(tf.argmax(logit_softmax, 1), y_input), tf.float32))

 saver = tf.train.Saver()
 config = tf.ConfigProto()
 config.allow_soft_placement = True
 config.gpu_options.allow_growth = True

  exporter = structure_exporter.StructureExporter(network_regularizer.op_regularizer_manager)
  now_lr = 0.1
  with tf.Session(config=config) as sess:
      sess.run(tf.global_variables_initializer())
      sess.run(tf.local_variables_initializer())
      counter = 0
      for i in range(epoch):
          sess.run(iterator.initializer)
         while True:
             try:
                 batch_train, label_train = sess.run(next_element)
                 _, loss_val,regloss_val, acc_val, lr_val,structure_exporter_tensors= sess.run([train_op, loss,regularizer_loss, acc, lr,exporter.tensors], feed_    dict={x_input: batch_train, y_input: label_train, lr: now_lr})
 
                 counter += 1
                 if counter % 100 == 0:
                     print('counter: ', counter, 'loss_val', loss_val,'regloss_val', regloss_val, 'acc: ', acc_val)
                 if counter % 1000 == 0:
                     exporter.populate_tensor_values(structure_exporter_tensors)
                     exporter.create_file_and_save_alive_counts(alive_dir, counter)
                     print('start test ')
                    sess.run(iterator_test.initializer)
                    avg_acc = []
`
The code in resnet50.py is as follows:
`
def identity_block2d(input_tensor, kernel_size, filters, stage, block, is_training, reuse, kernel_initializer=None):
	filters1, filters2, filters3 = filters

	conv_name_1 = 'conv' + str(stage) + '_' + str(block) + '_1x1_reduce'
	bn_name_1   = 'bn'   + str(stage) + '_' + str(block) + '_1x1_reduce'

	x = tf.layers.conv2d(input_tensor, filters1, (1, 1), use_bias=False, padding='SAME', name=conv_name_1, reuse=reuse, kernel_initializer=kernel_initializer)
	x = tf.layers.batch_normalization(x, training=is_training, name=bn_name_1, reuse=reuse)
	x = tf.nn.relu(x)

	conv_name_2 = 'conv' + str(stage) + '_' + str(block) + '_3x3'
	bn_name_2   = 'bn'   + str(stage) + '_' + str(block) + '_3x3'
	x = tf.layers.conv2d(x, filters2, kernel_size=(3,3), padding='SAME', use_bias=False, name=conv_name_2, reuse=reuse, kernel_initializer=kernel_initializer)
	x = tf.layers.batch_normalization(x, training=is_training, name=bn_name_2, reuse=reuse)
	x = tf.nn.relu(x)

	conv_name_3 = 'conv' + str(stage) + '_' + str(block) + '_1x1_increase'
	bn_name_3   = 'bn'   + str(stage) + '_' + str(block) + '_1x1_increase'
	x = tf.layers.conv2d(x, filters3, (1,1), name=conv_name_3, padding='SAME', use_bias=False, reuse=reuse, kernel_initializer=kernel_initializer)
	x = tf.layers.batch_normalization(x, training=is_training, name=bn_name_3, reuse=reuse)
	
	x = tf.add(input_tensor, x)
	x = tf.nn.relu(x)
	return x


def conv_block_2d(input_tensor, kernel_size, filters, stage, block, is_training, reuse, strides=(2, 2), kernel_initializer=None):
	filters1, filters2, filters3 = filters

	conv_name_1 = 'conv' + str(stage) + '_' + str(block) + '_1x1_reduce'
	bn_name_1   = 'bn'   + str(stage) + '_' + str(block) + '_1x1_reduce'
	x = tf.layers.conv2d(input_tensor, filters1, (1, 1), use_bias=False, padding='SAME',  strides=strides, name=conv_name_1, reuse=reuse, kernel_initializer=kernel_initializer)
	x = tf.layers.batch_normalization(x, training=is_training, name=bn_name_1, reuse=reuse)
	x = tf.nn.relu(x)

	conv_name_2 = 'conv' + str(stage) + '_' + str(block) + '_3x3'
	bn_name_2   = 'bn'   + str(stage) + '_' + str(block) + '_3x3'
	x = tf.layers.conv2d(x, filters2, kernel_size=(3,3), padding='SAME', use_bias=False, name=conv_name_2, reuse=reuse, kernel_initializer=kernel_initializer)
	x = tf.layers.batch_normalization(x, training=is_training, name=bn_name_2, reuse=reuse)
	x = tf.nn.relu(x)

	conv_name_3 = 'conv' + str(stage) + '_' + str(block) + '_1x1_increase'
	bn_name_3   = 'bn'   + str(stage) + '_' + str(block) + '_1x1_increase'
	x = tf.layers.conv2d(x, filters3, (1,1), name=conv_name_3, padding='SAME', use_bias=False, reuse=reuse, kernel_initializer=kernel_initializer)
	x = tf.layers.batch_normalization(x, training=is_training, name=bn_name_3, reuse=reuse)

	conv_name_4 = 'conv' + str(stage) + '_' + str(block) + '_1x1_shortcut'
	bn_name_4   = 'bn'   + str(stage) + '_' + str(block) + '_1x1_shortcut'
	shortcut = tf.layers.conv2d(input_tensor, filters3, (1,1), padding='SAME', use_bias=False, strides=strides, name=conv_name_4, reuse=reuse, kernel_initializer=kernel_initializer)
	shortcut = tf.layers.batch_normalization(shortcut, training=is_training, name=bn_name_4, reuse=reuse)

	x = tf.add(shortcut, x)
	x = tf.nn.relu(x)
	return x


def resnet50(input_tensor, is_training=True, pooling_and_fc=True, reuse=False, kernel_initializer=tf.contrib.layers.xavier_initializer()):
	x = tf.layers.conv2d(input_tensor, 64, (3,3), strides=(1,1), padding='SAME', use_bias=False, kernel_initializer=kernel_initializer, name='face_conv1_1/3x3_s1', reuse=reuse)
	x = tf.layers.batch_normalization(x, training=is_training, name='face_bn1_1/3x3_s1', reuse=reuse)
	x = tf.nn.relu(x)
	# x = tf.layers.max_pooling2d(x, (2,2), strides=(2,2), name='mpool1')

	x1 = conv_block_2d(x, 3, [64, 64, 256], stage=2, block='face_1a', strides=(1,1), is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x1 = identity_block2d(x1, 3, [64, 64, 256], stage=2, block='face_1b', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x1 = identity_block2d(x1, 3, [64, 64, 256], stage=2, block='face_1c', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)

	x2 = conv_block_2d(x1, 3, [128, 128, 512], stage=3, block='face_2a', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x2 = identity_block2d(x2, 3, [128, 128, 512], stage=3, block='face_2b', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x2 = identity_block2d(x2, 3, [128, 128, 512], stage=3, block='face_2c', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x2 = identity_block2d(x2, 3, [128, 128, 512], stage=3, block='face_2d', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)

	x3 = conv_block_2d(x2, 3, [256, 256, 1024], stage=4, block='face_3a' , is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x3 = identity_block2d(x3, 3, [256, 256, 1024], stage=4, block='face_3b', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x3 = identity_block2d(x3, 3, [256, 256, 1024], stage=4, block='face_3c', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x3 = identity_block2d(x3, 3, [256, 256, 1024], stage=4, block='face_3d', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x3 = identity_block2d(x3, 3, [256, 256, 1024], stage=4, block='face_3e', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x3 = identity_block2d(x3, 3, [256, 256, 1024], stage=4, block='face_3f', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)


	x4 = conv_block_2d(x3, 3, [512, 512, 2048], stage=5, block='face_4a', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x4 = identity_block2d(x4, 3, [512, 512, 2048], stage=5, block='face_4b', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)
	x4 = identity_block2d(x4, 3, [512, 512, 2048], stage=5, block='face_4c', is_training=is_training, reuse=reuse, kernel_initializer=kernel_initializer)

	if pooling_and_fc:
		# pooling_output = tf.layers.max_pooling2d(x4, (7,7), strides=(1,1), name='mpool2')
		print('before gap: ', x4)
		pooling_output = tf.reduce_mean(x4, [1, 2])
		fc_output      = tf.layers.dense(pooling_output, 100, name='face_fc1', reuse=reuse, kernel_initializer=tf.contrib.layers.xavier_initializer(), bias_initializer=tf.zeros_initializer())
		# fc_output      = tf.layers.batch_normalization(fc_output, training=is_training, reuse=reuse, name='face_fbn')

	return fc_output
`

**Comment by eladeban** [comment]
Did you get it to work?

**Comment by H-arshit** [comment]
Well am getting the same issue..
I am trying to use Model Size Regularizer and my regularization_loss turns out to be nan..

**Comment by eladeban** [comment]
if you have a specific question or problem, please reopen
